### PR TITLE
[image] Add animator for transitioning web images

### PR DESCRIPTION
**Pull request by aleqsio**
# Why
This PR allows us to load in images only after they have been loaded and with transitions between them.

# How

By storing children of `AnimationManager` in state and keeping them rendered in until after they fire an [onanimationend](https://developer.mozilla.org/en-US/docs/Web/API/Element/animationend_event) event.

If you set the `initial` prop this will be the first node mounted (aimed for placeholders or blurhash).

Animation is done through adding and removing CSS classes

# Test Plan

Manually tested

https://user-images.githubusercontent.com/5597580/210216330-ffca361a-4a84-4c36-9345-50d8ea640e63.mov

https://user-images.githubusercontent.com/5597580/210216379-c12f7d3a-9880-455c-a0f3-987ba202ca41.mov

https://user-images.githubusercontent.com/5597580/210216447-208c5336-6fb7-4449-8e59-cd20aa4034f5.mov
note the delay on setting source

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-image/src/ExpoImage.web.tsx [hunk 1]
```diff
@@ -280,6 +31,19 @@ function onErrorAdapter(onError?: { (event: { error: string }): void }) {
   };
 }
 
+const SUPPORTED_ANIMATIONS = [
```
Comment (tsapeta) [inline packages/expo-image/src/ExpoImage.web.tsx:34]: Probably worth adding a type here
```suggestion
const SUPPORTED_ANIMATIONS: ImageTransition['effect'][] = [
```

File: packages/expo-image/src/web/AnimationManager.tsx [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    // key, ReactElement, ref, callbacks
+    return { animationKey, child, ref, callbackContainer };
+  }, [node?.[1]]);
+  return newNode;
+}
+
+const validateTimingFunctionForAnimation = (
+  animationClass: string | null,
+  timingFunction: string | null
+) => {
```
Comment (tsapeta) [inline packages/expo-image/src/web/AnimationManager.tsx:65]:
```suggestion
function validateTimingFunctionForAnimation(
  animationClass: string | null,
  timingFunction: string | null
) {
```

I guess the `timingFunction` could be of `ImageTransition['timing']` type

File: packages/expo-image/src/web/AnimationManager.tsx [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            n.map((n2) => n2.ref)
+          );
+        } else {
+          n.forEach((oldNode) => oldNode.callbackContainer?.onAnimationFinished?.(true));
+        }
+      };
+      n.forEach((prevNode) => (prevNode.callbackContainer.onReady = () => null));
+      return [...n, newNode];
+    });
+  }, [newNode]);
```
Comment (tsapeta) [inline packages/expo-image/src/web/AnimationManager.tsx:188]: Shouldn't `animation` be in the dependencies too?

File: packages/expo-image/src/web/style.tsx [hunk 4]
```diff
@@ -0,0 +1,121 @@
+const style = `
+        .cross-dissolve {
```
Comment (tsapeta) [inline packages/expo-image/src/web/style.tsx:2]: Kinda weird indentation 😅 I would say no indent for selectors is the best choice

File: packages/expo-image/src/web/style.tsx [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        }
+        .flip-from-bottom-end {
+          transform:  translateZ(calc(var(--expo-image-height,1000px) * -1.25)) rotateX(180deg);
+          opacity: 0;
+        }
+          `;
+export default function loadStyle() {
+  const styleTag = document.createElement('style');
+  styleTag.innerHTML = style;
+  document.head.appendChild(styleTag);
```
Comment (tsapeta) [inline packages/expo-image/src/web/style.tsx:120]: So we cannot scope these styles and put inside the image container? The problem here is that these class names are a bit generic and the user may already have some elements that would be affected by these styles. Maybe at least prefix them with `expo` or `expo-image`?

File: packages/expo-image/src/web/AnimationManager.tsx (hunk 3 above)
Comment (aleqsio) [reply packages/expo-image/src/web/AnimationManager.tsx]: Fixed by moving animation elsewhere

File: packages/expo-image/src/web/style.tsx (hunk 5 above)
Comment (aleqsio) [reply packages/expo-image/src/web/style.tsx:132]: Yes, added a descendant selector

File: packages/expo-image/src/web/AnimationManager.tsx [hunk 6]
```diff
@@ -53,42 +54,62 @@ function validateTimingFunctionForAnimation(
...
-  timingFunction: ImageTransition['timing']
-) {
-  if (!animationClass) return null;
-  const timingClass = `image-timing-${validateTimingFunctionForAnimation(
-    animationClass,
-    timingFunction
-  )}`;
+function validateAnimationClass(effect: ImageTransition['effect']) {
+  if (!effect) return null;
+  if (SUPPORTED_ANIMATIONS.includes(effect)) return effect;
```
Comment (tsapeta) [inline packages/expo-image/src/web/AnimationManager.tsx:59]:
```suggestion
  if (!effect) {
    return null;
  }
  if (SUPPORTED_ANIMATIONS.includes(effect)) {
    return effect;
  }
```

File: packages/expo-image/src/web/AnimationManager.tsx [hunk 7]
```diff
@@ -53,42 +54,62 @@ function validateTimingFunctionForAnimation(
...
+  return 'cross-dissolve';
+}
+
+export function getAnimatorFromTransition(transition: ImageTransition | null | undefined) {
+  if (!transition?.duration) return null;
```
Comment (tsapeta) [inline packages/expo-image/src/web/AnimationManager.tsx:64]:
```suggestion
  if (!transition?.duration) {
    return null;
  }
```

**Review by tsapeta — APPROVED** [review]

**Pull request merged**